lychee-lib/src/types/basic_auth.rs

```diff
@@ -0,0 +1,174 @@
+use std::str::FromStr;
+
+use headers::{authorization::Basic, Authorization};
+use regex::Regex;
+use serde::Deserialize;
+use serde_with::DeserializeFromStr;
+use thiserror::Error;
+
+#[derive(Copy, Clone, Debug, Error)]
+pub enum BasicAuthCredentialsParseError {
+    #[error("Invalid basic auth credentials syntax")]
+    InvalidSyntax,
+
+    #[error("Missing basic auth password")]
+    MissingPassword,
+
+    #[error("Missing basic auth username")]
+    MissingUsername,
+
+    #[error("Too many values separated by colon. Expected 2, got {0}. Valid form is '<username>:<password>'")]
+    TooManyParts(usize),
+}
+
+/// [`BasicAuthCredentials`] contains a pair of basic auth values consisting of
+/// a username and password.
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+pub struct BasicAuthCredentials {
+    /// Basic auth username
+    pub username: String,
+
+    /// Basic auth password
+    pub password: String,
+}
+
+impl FromStr for BasicAuthCredentials {
+    type Err = BasicAuthCredentialsParseError;
+
+    fn from_str(credentials: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<_> = credentials.trim().split(':').collect();
+
+        if parts.len() == 1 {
+            return Err(BasicAuthCredentialsParseError::InvalidSyntax);
+        }
+
+        if parts.len() != 2 {
+            return Err(BasicAuthCredentialsParseError::TooManyParts(parts.len()));
+        }
+
+        if parts[0].is_empty() {
```

Comment:
A single space ` ` would not be empty. Would it be a valid basic auth username, though? Shall we apply `trim` before `is_empty`?